Fix Intro Offer price description

IanRamosC opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

Fixes the billing term text for 1201210512993648-as-1203748989040194

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
* Fix the discounted price and details text for pricing cards that contains a product that we offer a trial (an intro offer for the first month).
<img width="457" alt="image" src="https://user-images.githubusercontent.com/5714212/213015382-52793532-9cb7-484a-ad1d-3f10803d8d0f.png">
<img width="655" alt="image" src="https://user-images.githubusercontent.com/5714212/213015494-81114a40-229d-474c-aa44-d1e8451165f0.png">


### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
p1673966558676559-slack-C02LK1W8T4Z


## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Checkout this branch locally
* Run `jetpack build packages/backup` and `jetpack build packages/my-jetpack`
* Go to My Jetpack, then Add VaultPress Backup
* Make sure the right price is present ($1/€1)
* Make sure that the text says "for the first month, billed yearly" instead of "/month, billed yearly"
* Install VaultPress Backup
* Without a plan, go to `/wp-admin/admin.php?page=jetpack-backup`
* Confirm that the fix for both the price and details text is there

Note: if you use a currency other than USD or EUR, it should display the localized price, with the correct currency code.

IanRamosC committed:
Fix how to calc Jetpack Backup intro offer price

IanRamosC committed:
Add intro offer object to ProductDetailCard

IanRamosC committed:
changelog

IanRamosC committed:
Backup: fix PricingCard currency and details text

CodeyGuyDylan commented on projects/packages/backup/src/js/hooks/useConnection.js:
```diff
@@ -33,7 +33,15 @@ export default function useConnection() {
 			setPrice( res.cost / 12 );
 			if ( res.introductory_offer ) {
-				setPriceAfter( res.introductory_offer.cost_per_interval / 12 );
+				const {
```
Since this code is a replica of the other one checking for an intro discount, I wonder if we could make this an external function called "getDiscountedPrice" or something like that?

CodeyGuyDylan reviewed (changes requested):
I didn't spot anything wrong with the code itself, but it seems like quite a few tests are failing, it looks like most or all of them are for my-jetpack failing because it is not expecting the `introductory_offer`, so you'll need to update whatever test is checking for that

Additionally, there is some versioning errors with the changelog tests, I didn't dig into it much but perhaps a rebase would fix this? No idea but that'd be my first attempt to fix it 😅 

Other than this it all looks good other than one suggestion in the code

IanRamosC committed:
Products: only return intro offer when available

This commit fixes the return of Wpcom_Products::get_product_pricing to only return `introductory_offer` when available

IanRamosC committed:
Merge branch 'trunk' into update/intro-offer-price-display-and-description

IanRamosC committed:
Products: fix test_get_product_price

IanRamosC committed:
Fix version mismatch

robertsreberski commented on projects/packages/backup/src/js/components/Admin.js:
```diff
@@ -340,6 +353,11 @@ const NoBackupCapabilities = () => {
 		'jetpack-backup-pkg'
 	);
+	const priceDetails =
+		introOffer && introOffer.interval_unit === 'month'
```
Should we also check the `intervalCount === 1`?

robertsreberski commented on projects/packages/my-jetpack/_inc/components/product-detail-card/index.jsx:
```diff
@@ -145,6 +147,11 @@ const ProductDetailCard = ( { slug, onClick, trackButtonClick, className, suppor
 		: null;
 
+	const priceDescription =
+		isIntroductoryOffer && introductoryOffer.intervalUnit === 'month'
```
Same as above, should we check `intervalCount === 1`?

robertsreberski commented on projects/packages/my-jetpack/_inc/state/selectors.js:
```diff
@@ -10,14 +10,21 @@ export const getProduct = ( state, productId ) => {
 		Math.ceil( ( product.pricingForUi.fullPrice / 12 ) * 100 ) / 100;
+	const discountedPricePerMonth =
+		product.pricingForUi.isIntroductoryOffer &&
+		product.pricingForUi.introductoryOffer.intervalUnit === 'month'
```
Same here, should we take `intervalCount` into consideration in these checks? Let's be consistent either way.

robertsreberski reviewed (commented):
Left questions about either including `intervalCount === 1` check in the PR or not. See below

jboland88 commented on projects/packages/backup/src/js/components/Admin.js:
```diff
@@ -317,13 +317,26 @@ const NoBackupCapabilities = () => {
+					interval_unit: intervalUnit,
+				} = res.introductory_offer;
+				const discountedPrice =
+					intervalCount === 1 && intervalUnit === 'month' ? costPerInterval : costPerInterval / 12;
```
I wonder if we can go ahead and move towards a more flexible monthly price calculation function. 
Something like this maybe (pseudocode)

```js
// can pass intro offer or plan
function get_monthly_price( intro_offer ) {
    if ( intro_offer.intervalUnit === 'month' ){
        return intro_offer.costPerInterval / intro_offer.intervalCount
    } else if  ( intro_offer.intervalUnit === 'year' ) {
        return intro_offer.costPerInterval / ( intro_offer.intervalCount * 12 )
    }
}
```

That way we could handle things in a more robust way

IanRamosC committed:
Backup: revert intro offer calculation

IanRamosC committed:
My Jetpack: revert monthly price calculation

IanRamosC committed:
My Jetpack: fix price detail text condition

IanRamosC committed:
Backup: fix price detail text condition

CodeyGuyDylan commented on projects/packages/my-jetpack/_inc/components/product-detail-card/index.jsx:
```diff
@@ -145,6 +146,11 @@ const ProductDetailCard = ( { slug, onClick, trackButtonClick, className, suppor
 				} )
 		: null;
 
+	const priceDescription =
```
I still think this could be some sort of external function or util, but I know we are going to tackle the pricing at a much bigger scale soon so I think this is fine for now

CodeyGuyDylan reviewed (dismissed):
Other than a couple of failing tests that may have to be manually addressed, but don't seem to have to do with these specific changes, all looks good!

IanRamosC committed:
Fix changelogger version

IanRamosC committed:
Merge branch 'trunk' into update/intro-offer-price-display-and-description

IanRamosC committed:
Fix bad minification issue on production builds

coder-karen reviewed (approved):
Approving based on everything working well following the testing instructions 👍
